File: backend/src/main/java/com/project/util/WordFrequencyCounter.java
package com.project.util;

import org.springframework.stereotype.Component;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

@Component
public class WordFrequencyCounter {

    /**
     * Takes a list of words and uses a HashMap to count how many times each word appears.
     */
    public Map<String, Integer> countWordFrequencies(List<String> words) {
        
        // 1. Create our empty HashMap to act as a digital tally counter
        Map<String, Integer> frequencyMap = new HashMap<>();
        
        // 2. Loop through every single word in the list provided to us
        for (String word : words) {
            
            // Standardize the word to lowercase so "Apple" and "apple" get counted together
            String lowerCaseWord = word.toLowerCase();
            
            // 3. The Core Logic: Check if we have already seen this word before
            if (frequencyMap.containsKey(lowerCaseWord)) {
                // If we HAVE seen it: Grab its current total, add 1, and put it back
                int currentCount = frequencyMap.get(lowerCaseWord);
                frequencyMap.put(lowerCaseWord, currentCount + 1);
                
            } else {
                // If we have NEVER seen it: Add it to the map for the first time with a count of 1
                frequencyMap.put(lowerCaseWord, 1);
            }
        }
        
        // 4. Return the finished Map with all the final totals
        return frequencyMap;
    }
}

File: node-backend/utils/wordFrequencyCounter.js
import { cleanWord } from './wordCleaner.js';

/**
 * Takes an array of words and counts how many times each word appears (lowercased).
 *
 * @param {string[]} words
 * @returns {Record<string, number>} A map of word to frequency.
 */
export function countWordFrequencies(words) {
  const frequencyMap = {};

  if (!words || !Array.isArray(words)) {
    return frequencyMap;
  }

  for (const word of words) {
    const cleaned = cleanWord(word);
    
    if (cleaned) {
      if (frequencyMap[cleaned]) {
        frequencyMap[cleaned] += 1;
      } else {
        frequencyMap[cleaned] = 1;
      }
    }
  }

  return frequencyMap;
}
